### DISCUSSION TITLE
Multi module in PDM

### DISCUSSION THREAD

**Discussion by tisonkun** [pdm-project/pdm#2700] [Q&A]
Maven supports multiple modules and Cargo supports workspace.

I wonder if PDM has similar supports. I know we can have multiple Python modules in one project, but what if I'd like to write a monorepo with multiple projects? Is there some out-of-the-box supports or I just layout as:

```
/
|- project-a / pyproject.toml
|- project-b / pyproject.toml
...
```

I may expect

```
/
|- project-a / pyproject.toml
|- project-b / pyproject.toml
...
| - pyproject.toml
```

So that we can share some settings like metadata and dependencies management.

**Comment by rilshok** [comment]
I support the @tisonkun's request. This functionality is very important for me, I am actively looking for actual examples of how to properly build a mono repository consisting of many sub-modules.

**Comment by dranjan** [comment]
I haven't tried these yet, but I'll share what I know.

- This topic has been discussed here: https://github.com/pdm-project/pdm/issues/1505
- There's also some information about this in the documentation: https://pdm-project.org/en/stable/usage/advanced/#use-pdm-to-manage-a-monorepo

One thing I'm not completely sure about is the workflow for managing dependencies of the subprojects, but maybe this is viable:
1. Manually modify the `pyproject.toml` in one or more subprojects.
2. Run `pdm lock --update-reuse` at the top level to update `pdm.lock`. (With `--update-reuse`, I think the edit+lock sequence shouldn't be too far from using `pdm add|remove|update` directly. I don't see a way to use these latter commands in the monorepo setup.)

**Comment by aniaan** [comment]
@tisonkun You can give [uv](https://docs.astral.sh/uv/concepts/workspaces/) a try.
